### DISCUSSION TITLE
Notebook Mega Thread!

### DISCUSSION THREAD

**Discussion by znorman-harris** [interactive-data-language/vscode-idl#6] [Ideas]
Hey All! We have an early preview of notebooks for IDL as part of the extension.

This is a thread where you can ask questions, report problems, or provide feedback on how they behave.

**Comment by thepeteriley** [comment]
I love the Notebooks! I'm having issues generating object graphics in them though. I thought it was something in my notebook, but even running the example notebook, I get the following error when running the first plot cell:

% PLOT: Attempt to call undefined procedure: 'DEFAULT'.
% Execution halted at: $MAIN$    

The code in the cell is: 

; make some plots!
p1 = plot(/test)
p2 = surface(/test)
p4 = barplot(/test)

I get the same error with:

p1 = plot(findgen(100))

Direct graphics are fine, so PLOT,FINDGEN(100) works. 

-Pete

**Reply by znorman-harris** [reply]
Thanks for reaching out again! And that's an interesting error, I don't know if it is related to the extension or not.

Can you execute "p = plot(/test)" from an IDL session in VSCode without notebooks?

If so, I'll reach out to one of the team members who has an Arm Mac and see if he can reproduce it.

**Comment by thepeteriley** [comment]
Yes, p = plot(/test) works in 8.8.2 and 9.0 from either the command line or within the IDE.

**Reply by znorman-harris** [reply]
OK, thanks for checking! Let me see if I can get a reproduce case for this today and figure out what's going on.

**Reply by znorman-harris** [reply]
A couple of questions:

1. If you disable the setting to embed graphics, do the notebooks now work? I know it's not really a "notebook" at that point, but it might be a temporary workaround. You can find this setting here:

![image](https://github.com/interactive-data-language/vscode-idl/assets/31664668/50a2440f-0935-40d6-b015-fef48cc71aa4)

2. Are you on the latest version of VSCode (1.85)?
3. When you first run a notebook cell, and you see the "starting IDL" message appear, what do your logs show? From the sidebar, you can select "View Extension Logs" all the way at the bottom. You should see something like this:

```
notebook-idl info Starting IDL
extensionHostProcess.js:147
   {
    cmd: 'C:\\Program Files\\NV5\\ENVI60\\IDL90\\bin\\bin.x86_64\\idl',
    dir: 'C:\\Program Files\\NV5\\ENVI60\\IDL90',
    path: '+C:\\Users\\znorman\\Documents\\node\\vscode-idl-true\\idl\\vscode;+C:\\Users\\znorman\\Documents\\node\\vscode-idl-true\\idl\\test\\scratch;<IDL_DEFAULT>',
    dlm_path: 'C:\\Program Files\\NV5\\ENVI60\\IDL90\\bin\\bin.x86_64'
  }
```

I'm looking to verify the path specifically.

**Comment by thepeteriley** [comment]
Hi Zach, 

For 1., unfortunately, this doesn't resolve it. The code pops open an X graphics window now, but the error is the same:

% PLOT: Attempt to call undefined procedure: 'DEFAULT'.
% Execution halted at: $MAIN$             

For 2., Yes, I'm using 1.85.1

For 3, there's lots of info int the Extension Logs file, but no string with "Starting IDL"...just: 

idl-client info Starting the language server using "VSCode". Output from "node --version" (minimum is "16.0.0"):
  Command failed: node --version
  /bin/sh: node: command not found
  
idl-client info Starting language server
idl-client info Registering debug configuration provider
idl-client info Starting IDL debug server (not IDL)

However, if I run a cell with just "PLOT,FINDGEN(10)" that renders fine, so I think the server is running. 

Looking for "NV5" as another clue, at the top is:

idl-client info IDL directory and path: "/Applications/NV5/idl90/bin/bin.darwin.arm64"

and later: 

    debugMode: false,
    IDL: {
      directory: '/Applications/NV5/idl90/bin/bin.darwin.arm64',

-Pete

**Reply by znorman-harris** [reply]
Ah! It looks like the notebook log (and a few others) were just showing up for our development environment and not when it is officially installed. I just fixed this and, if you update to the latest `4.2.1`, then the logs should appear now.

Here's a video showing how to get the logs:

https://github.com/interactive-data-language/vscode-idl/assets/31664668/5ddb3c04-4a36-4427-b0d5-44d9dbb1fe70

One more question: do plots have the same issue from a debug session? You can launch that from the icon in the lower left of the status bar.

Edit: Do you get a different error message when you use "p = plot(/test, /debug)"?

**Reply by thepeteriley** [reply]
Hi Zach, 

TL;DR: It turned out to be a conflict with a function "uniq.pro" for which I had my own version. I renamed this and everything works as advertised. 

Now the longer version....

I installed 4.2.1 and reran, getting:

idl-notebook-session info Starting IDL
  {
    cmd: '/Applications/NV5/idl90/bin/bin.darwin.arm64/idl',
    dir: '/Applications/NV5/idl90',
    path: '+/Users/pete/.vscode/extensions/idl.idl-for-vscode-4.2.1/idl/vscode:+:+/Users/pete/Dropbox/idl/lib/:<IDL_DEFAULT>',
    dlm_path: '/Applications/NV5/idl90/bin/bin.darwin.arm64'
  }
idl-notebook-session info IDL has started!
idl-command info View preferences

So far so good.

I did get the same error with the /debug parameter, but now, there's a lot more information. In fact, it had the solution: 

% Attempt to call undefined procedure: 'DEFAULT'.
% Execution halted at: UNIQ               82
   [/Users/pete/Dropbox/idl/lib/idl_lib_ssw/uniq.pro](https://file+.vscode-resource.vscode-cdn.net/Users/pete/Dropbox/idl/lib/idl_lib_ssw/uniq.pro)
%                      IDLITPARAMETER::GETPARAMETERTYPES  631
   [/Applications/NV5/idl90/lib/itools/framework/idlitparameter__define.pro](https://file+.vscode-resource.vscode-cdn.net/Applications/NV5/idl90/lib/itools/framework/idlitparameter__define.pro)
%                      IDLITOBJDESCVIS::GETDATATYPES   75
   [/Applications/NV5/idl90/lib/itools/framework/idlitobjdescvis__define.pro](https://file+.vscode-resource.vscode-cdn.net/Applications/NV5/idl90/lib/itools/framework/idlitobjdescvis__define.pro)
%                      IDLITSRVCREATEVISUALIZATION::_CHECKVISDATASUPPORT 1005
   [/Applications/NV5/idl90/lib/itools/framework/idlitsrvcreatevisualization__](https://file+.vscode-resource.vscode-cdn.net/Applications/NV5/idl90/lib/itools/framework/idlitsrvcreatevisualization__)
  define.pro
%                      IDLITSRVCREATEVISUALIZATION::CREATEVISUALIZATION 1247
   [/Applications/NV5/idl90/lib/itools/framework/idlitsrvcreatevisualization__](https://file+.vscode-resource.vscode-cdn.net/Applications/NV5/idl90/lib/itools/framework/idlitsrvcreatevisualization__)
  define.pro
%                      IDLITSYSTEM::CREATEVISUALIZATION 2769
   [/Applications/NV5/idl90/lib/itools/framework/idlitsystem__define.pro](https://file+.vscode-resource.vscode-cdn.net/Applications/NV5/idl90/lib/itools/framework/idlitsystem__define.pro)
%                      IDLITSYSTEM::CREATETOOL 2526
   [/Applications/NV5/idl90/lib/itools/framework/idlitsystem__define.pro](https://file+.vscode-resource.vscode-cdn.net/Applications/NV5/idl90/lib/itools/framework/idlitsystem__define.pro)
%                      IDLITSYS_CREATETOOL  419
   [/Applications/NV5/idl90/lib/itools/framework/idlitsys_createtool.pro](https://file+.vscode-resource.vscode-cdn.net/Applications/NV5/idl90/lib/itools/framework/idlitsys_createtool.pro)
%                      IPLOT             512
   [/Applications/NV5/idl90/lib/itools/iplot.pro](https://file+.vscode-resource.vscode-cdn.net/Applications/NV5/idl90/lib/itools/iplot.pro)
%                      GRAPHIC           416
   [/Applications/NV5/idl90/lib/graphics/graphic.pro](https://file+.vscode-resource.vscode-cdn.net/Applications/NV5/idl90/lib/graphics/graphic.pro)
%                      PLOT               83

So, the uniq.pro in a 3rd party library I use must have been conflicting with default.pro. I renamed this and ran again. Worked like a charm! I put it back in embed mode, and the images appeared nicely inline. 

Thanks for your help! Sorry, it turned out to be a "me" problem. 

Pete

**Reply by znorman-harris** [reply]
Ah, that will do it! We do have our own IDL function called uniq (and it ships with IDL), so might be worth keeping that in mind in the future.

And no worries about it being something you had set up. I have what my good friend calls "Zach Problems" all the time when it comes to code, just glad we got it sorted out!

**Comment by vasu-gupta01** [comment]
Hi,

I am having an issue with embedded graphics in the included IDL Notebook Example. When I run the the Graphics cell, a blank plot window shows up and the following error is thrown:

![image](https://github.com/interactive-data-language/vscode-idl/assets/50789491/a78b71a6-5106-4307-9468-14a10555dd96)

Output log:

![image](https://github.com/interactive-data-language/vscode-idl/assets/50789491/737647ae-e652-44e5-a497-536a0f20974c)


Note: I have enabled the 'Embed Graphics' setting.

How can I resolve this issue?

Vasu

**Reply by znorman-harris** [reply]
Hey There! Sorry it took a few days to get back to you, I took a few days off and am catching up.

What version of IDL are you using? The system variable we need for notebooks was added in the last couple of years (I need to check to be sure when) and requires a newer version of IDL.

**Reply by vasu-gupta01** [reply]
Thanks for the reply! I am currently using IDL 8.3 on Windows

**Reply by znorman-harris** [reply]
Just checked with the IDL team, you'll need a newer version of IDL in order for notebooks to work right. You're using a version before the python bridge (and support for Jupyter notebooks) which requires the `!magic` system variable in order to function properly.

In our next release we will have a check when notebooks start up to make sure you have a newer version so users don't get the weird errors like you have encountered.

**Comment by fermindav1d** [comment]
I have an issue with the notebooks (and IDL for Vscode overall), where I can't run IDL at all. I am new to IDL in general, but I think the IDL notebooks would be useful for my summer project. Currently, I am trying to run this on NERSC Perlmutter, which I connect to via SSH. I can run IDL from the command line there, but I can't start an IDL session on VScode. How do I resolve this issue?

**Reply by znorman-harris** [reply]
Hey There!

Right now, when you try to start IDL through VSCode, it is going to look for a local installation of IDL on the same machine that VSCode is running on.

This means that, if you are connecting via SSH and running IDL in a terminal via SSH, that VSCode has no connectivity to the IDL session.

If you need to access IDL on another machine, you might need to play around with remote development with VSCode. Maybe something from here? https://code.visualstudio.com/docs/remote/remote-overview

**Reply by fermindav1d** [reply]
I will give it a try! Thank you.
